File: Frontend/src/lib/services/likedSongs.ts
import { supabase } from '../supabaseClient';

export interface LikedSong {
  id: string;
  user_id: string;
  song_id: string;
  song_name: string;
  artist_name: string;
  album_name?: string;
  image_url?: string;
  spotify_url?: string;
  duration_ms?: number;
  liked_at: string;
}

/**
 * Service for managing user's liked songs
 */
export class LikedSongsService {
  private static instance: LikedSongsService;
  private likedSongsCache: Set<string> = new Set();
  private cacheInitialized: boolean = false;

  private constructor() {}

  static getInstance(): LikedSongsService {
    if (!LikedSongsService.instance) {
      LikedSongsService.instance = new LikedSongsService();
    }
    return LikedSongsService.instance;
  }

  /**
   * Initialize cache with user's liked songs
   */
  async initializeCache(): Promise<void> {
    try {
      const { data: { session } } = await supabase.auth.getSession();
      if (!session) {
        this.cacheInitialized = true;
        return;
      }

      const token = session.access_token;
      const apiUrl = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000';
      
      const response = await fetch(`${apiUrl}/api/liked-songs`, {
        headers: {
          'Authorization': `Bearer ${token}`,
        },
      });

      if (response.ok) {
        const likedSongs: LikedSong[] = await response.json();
        this.likedSongsCache = new Set(likedSongs.map(song => song.song_id));
        console.log(`✅ Loaded ${this.likedSongsCache.size} liked songs into cache`);
      }
      
      this.cacheInitialized = true;
    } catch (error) {
      console.error('Failed to initialize liked songs cache:', error);
      this.cacheInitialized = true;
    }
  }

  /**
   * Get all liked songs
   */
  async getLikedSongs(limit: number = 50): Promise<LikedSong[]> {
    try {
      const { data: { session } } = await supabase.auth.getSession();
      if (!session) {
        console.log('No session, returning empty liked songs');
        return [];
      }

      const token = session.access_token;
      const apiUrl = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000';
      
      const response = await fetch(`${apiUrl}/api/liked-songs?limit=${limit}`, {
        headers: {
          'Authorization': `Bearer ${token}`,
        },
      });

      if (!response.ok) {
        throw new Error(`Failed to fetch liked songs: ${response.statusText}`);
      }

      const likedSongs: LikedSong[] = await response.json();
      return likedSongs;
    } catch (error) {
      console.error('Error fetching liked songs:', error);
      return [];
    }
  }

  /**
   * Like a song
   */
  async likeSong(
    songId: string,
    songName: string,
    artistName: string,
    albumName?: string,
    imageUrl?: string,
    spotifyUrl?: string,
    durationMs?: number
  ): Promise<boolean> {
    try {
      const { data: { session } } = await supabase.auth.getSession();
      if (!session) {
        console.log('No session, cannot like song');
        return false;
      }

      const token = session.access_token;
      const apiUrl = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000';
      
      const formData = new FormData();
      formData.append('song_id', songId);
      formData.append('song_name', songName);
      formData.append('artist_name', artistName);
      if (albumName) formData.append('album_name', albumName);
      if (imageUrl) formData.append('image_url', imageUrl);
      if (spotifyUrl) formData.append('spotify_url', spotifyUrl);
      if (durationMs) formData.append('duration_ms', durationMs.toString());

      const response = await fetch(`${apiUrl}/api/liked-songs`, {
        method: 'POST',
        headers: {
          'Authorization': `Bearer ${token}`,
        },
        body: formData,
      });

      if (!response.ok) {
        throw new Error(`Failed to like song: ${response.statusText}`);
      }

      const result = await response.json();
      
      // Update cache
      this.likedSongsCache.add(songId);
      
      // Dispatch event to notify UI
      window.dispatchEvent(new CustomEvent('songLiked', { detail: { songId, liked: true } }));
      
      console.log(`❤️ Song liked: ${songName}`);
      return true;
    } catch (error) {
      console.error('Error liking song:', error);
      return false;
    }
  }

  /**
   * Unlike a song
   */
  async unlikeSong(songId: string): Promise<boolean> {
    try {
      const { data: { session } } = await supabase.auth.getSession();
      if (!session) {
        console.log('No session, cannot unlike song');
        return false;
      }

      const token = session.access_token;
      const apiUrl = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000';
      
      const response = await fetch(`${apiUrl}/api/liked-songs/${songId}`, {
        method: 'DELETE',
        headers: {
          'Authorization': `Bearer ${token}`,
        },
      });

      if (!response.ok) {
        throw new Error(`Failed to unlike song: ${response.statusText}`);
      }

      // Update cache
      this.likedSongsCache.delete(songId);
      
      // Dispatch event to notify UI
      window.dispatchEvent(new CustomEvent('songLiked', { detail: { songId, liked: false } }));
      
      console.log(`💔 Song unliked: ${songId}`);
      return true;
    } catch (error) {
      console.error('Error unliking song:', error);
      return false;
    }
  }

  /**
   * Check if a song is liked (uses cache if available)
   */
  async isLiked(songId: string): Promise<boolean> {
    // Initialize cache if not done
    if (!this.cacheInitialized) {
      await this.initializeCache();
    }

    // Check cache first
    if (this.cacheInitialized) {
      return this.likedSongsCache.has(songId);
    }

    // Fallback to API call
    try {
      const { data: { session } } = await supabase.auth.getSession();
      if (!session) {
        return false;
      }

      const token = session.access_token;
      const apiUrl = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000';
      
      const response = await fetch(`${apiUrl}/api/liked-songs/check/${songId}`, {
        headers: {
          'Authorization': `Bearer ${token}`,
        },
      });

      if (!response.ok) {
        return false;
      }

      const result = await response.json();
      return result.is_liked || false;
    } catch (error) {
      console.error('Error checking liked status:', error);
      return false;
    }
  }

  /**
   * Toggle like status of a song
   */
  async toggleLike(
    songId: string,
    songName: string,
    artistName: string,
    albumName?: string,
    imageUrl?: string,
    spotifyUrl?: string,
    durationMs?: number
  ): Promise<boolean> {
    const isCurrentlyLiked = await this.isLiked(songId);
    
    if (isCurrentlyLiked) {
      return await this.unlikeSong(songId);
    } else {
      return await this.likeSong(
        songId,
        songName,
        artistName,
        albumName,
        imageUrl,
        spotifyUrl,
        durationMs
      );
    }
  }

  /**
   * Clear cache (useful on logout)
   */
  clearCache(): void {
    this.likedSongsCache.clear();
    this.cacheInitialized = false;
  }
}

File: Backend/routes/liked_songs.py
"""
Liked Songs Routes
Handles user's liked songs with Prisma ORM
"""
from fastapi import APIRouter, HTTPException, Depends, Form
from typing import List, Optional
from pydantic import BaseModel
from datetime import datetime
import logging
from middleware.supabase_auth import get_current_user
from database import db

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/liked-songs", tags=["Liked Songs"])


class LikedSong(BaseModel):
    """Liked song model"""
    id: str
    user_id: str
    song_id: str
    song_name: str
    artist_name: str
    album_name: Optional[str] = None
    image_url: Optional[str] = None
    spotify_url: Optional[str] = None
    duration_ms: Optional[int] = None
    liked_at: str


class LikedSongResponse(BaseModel):
    """Response model for liked songs"""
    song_id: str
    song_name: str
    artist_name: str
    album_name: Optional[str] = None
    image_url: Optional[str] = None
    spotify_url: Optional[str] = None
    duration_ms: Optional[int] = None
    liked_at: str


@router.get("", response_model=List[LikedSongResponse])
async def get_liked_songs(
    limit: int = 50,
    user_data: dict = Depends(get_current_user)
):
    """
    Get user's liked songs using Prisma
    """
    try:
        user_id = user_data.get('sub')
        
        # Query liked songs using Prisma
        liked_songs = await db.likedsong.find_many(
            where={'user_id': user_id},
            order={'liked_at': 'desc'},
            take=limit
        )
        
        if not liked_songs:
            return []
        
        # Convert to response format
        return [
            LikedSongResponse(
                song_id=song.song_id,
                song_name=song.song_name,
                artist_name=song.artist_name,
                album_name=song.album_name,
                image_url=song.image_url,
                spotify_url=song.spotify_url,
                duration_ms=song.duration_ms,
                liked_at=song.liked_at.isoformat()
            )
            for song in liked_songs
        ]
        
    except Exception as e:
        logger.error(f"❌ Error fetching liked songs: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Failed to fetch liked songs: {str(e)}")


@router.post("")
async def like_song(
    song_id: str = Form(...),
    song_name: str = Form(...),
    artist_name: str = Form(...),
    album_name: Optional[str] = Form(None),
    image_url: Optional[str] = Form(None),
    spotify_url: Optional[str] = Form(None),
    duration_ms: Optional[int] = Form(None),
    user_data: dict = Depends(get_current_user)
):
    """
    Add a song to liked songs using Prisma
    """
    try:
        user_id = user_data.get('sub')
        
        # Check if already liked
        existing = await db.likedsong.find_first(
            where={
                'user_id': user_id,
                'song_id': song_id
            }
        )
        
        if existing:
            return {
                "success": True,
                "message": "Song already in liked songs",
                "already_liked": True
            }
        
        # Insert liked song
        await db.likedsong.create(
            data={
                'user_id': user_id,
                'song_id': song_id,
                'song_name': song_name,
                'artist_name': artist_name,
                'album_name': album_name,
                'image_url': image_url,
                'spotify_url': spotify_url,
                'duration_ms': duration_ms
            }
        )
        
        return {
            "success": True,
            "message": "Song liked successfully",
            "already_liked": False
        }
        
    except Exception as e:
        logger.error(f"❌ Error liking song: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Failed to like song: {str(e)}")


@router.delete("/{song_id}")
async def unlike_song(
    song_id: str,
    user_data: dict = Depends(get_current_user)
):
    """
    Remove a song from liked songs using Prisma
    """
    try:
        user_id = user_data.get('sub')
        
        # Delete liked song
        await db.likedsong.delete_many(
            where={
                'user_id': user_id,
                'song_id': song_id
            }
        )
        
        return {
            "success": True,
            "message": "Song unliked successfully"
        }
        
    except Exception as e:
        logger.error(f"❌ Error unliking song: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Failed to unlike song: {str(e)}")


@router.get("/check/{song_id}")
async def check_if_liked(
    song_id: str,
    user_data: dict = Depends(get_current_user)
):
    """
    Check if a song is liked by the user using Prisma
    """
    try:
        user_id = user_data.get('sub')
        
        liked_song = await db.likedsong.find_first(
            where={
                'user_id': user_id,
                'song_id': song_id
            }
        )
        
        is_liked = liked_song is not None
        
        return {
            "song_id": song_id,
            "is_liked": is_liked
        }
        
    except Exception as e:
        logger.error(f"❌ Error checking liked status: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Failed to check liked status: {str(e)}")
